# Introduction 

In classical programming, we feed the computer the rules and the data and it "computes" the answers.

Whereas, in machine learning we feed the computer some data and the expected results and it tries to determine a set of rules to achieve those results.

## A brief history of machine learning

1. Probabilistic modeling

2. Early neural networks

3. Kernel methods

4. Decision trees, random forests, and gradient boosting machines

5. Back to neural networks

## First example 

A network to categorize handwritten figure images.

First, we will have to prepare the libraries and datasets. 
The MNIST dataset comes preloaded in Keras, in the form of a set of four Numpy arrays.

In [8]:
from keras.datasets import mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

The workflow will be as follows: First, we’ll feed the neural network the training data, train_images and train_labels. The network will then learn to associate images and labels. Finally, we’ll ask the network to produce predictions for test_images, and we’ll verify whether these predictions match the labels from test_labels.

The core building block of neural networks is the layer, a data-processing module that you can think of as a filter for data. Some data goes in, and it comes out in a more use- ful form.

In [9]:
from keras import models
from keras import layers

network = models.Sequential()
network.add(layers.Dense(512, activation='relu', input_shape=(28 * 28,)))
network.add(layers.Dense(10, activation='softmax'))

As part of the compilation step, we need to pick three more things. 
* A loss function
* An optimizer
* Metrics to monitor during training and testing

In [11]:
network.compile(optimizer='rmsprop',
                loss='categorical_crossentropy',
                metrics=['accuracy'])

We can now prepare the image data.

In [12]:
train_images = train_images.reshape((60000, 28 * 28))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype('float32') / 255

We also need to categorically encode the labels

In [13]:
from keras.utils import to_categorical

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

We’re now ready to train the network, which in Keras is done via a call to the net- work’s fit method—we fit the model to its training data:

In [14]:
network.fit(train_images, train_labels, epochs=5, batch_size=128)

Epoch 1/5
60000/60000 [==============================] - 5s 82us/step - loss: 0.2599 - acc: 0.9254
Epoch 2/5
60000/60000 [==============================] - 5s 79us/step - loss: 0.1050 - acc: 0.9689
Epoch 3/5
60000/60000 [==============================] - 5s 84us/step - loss: 0.0703 - acc: 0.9789
Epoch 4/5
60000/60000 [==============================] - 5s 77us/step - loss: 0.0516 - acc: 0.9839
Epoch 5/5
60000/60000 [==============================] - 5s 77us/step - loss: 0.0385 - acc: 0.9885


Now, let's check how the model performs on the test.

In [18]:
test_loss, test_acc = network.evaluate(test_images, test_labels)

10000/10000 [==============================] - 1s 57us/step


In [19]:
print('test_acc:', test_acc)

test_acc: 0.9797


The accuracy is 97.97%. This concludes our first example.